### PR TITLE
Refactor callMain function in start.zig

### PR DESCRIPTION
**Pull request by gooncreeper**
Improves callMain in start.zig by switching on the main function's return type instead of doing a novemdecillion comparisons on the return type info.

### CODE REVIEW

File: lib/std/start.zig [hunk 1]
```diff
@@ -496,41 +496,44 @@ fn mainWithoutEnv(c_argc: c_int, c_argv: [*][*:0]c_char) callconv(.C) c_int {
...
-            switch (@typeInfo(@TypeOf(result))) {
-                .Void => return 0,
-                .Int => |info| {
-                    if (info.bits != 8 or info.signedness == .signed) {
-                        @compileError(bad_main_ret);
-                    }
-                    return result;
+        else => {
+            const return_info = @typeInfo(ReturnType);
+            if (comptime std.meta.activeTag(return_info) != .ErrorUnion) @compileError(bad_main_ret);
```
Comment (rohlem) [inline lib/std/start.zig:524]: tagged union instances can be directly compared to their enum tag types: `return_info != .ErrorUnion`

Comment (gooncreeper) [reply lib/std/start.zig]: Thanks! Fixed

File: lib/std/start.zig [hunk 2]
```diff
@@ -496,41 +496,44 @@ fn mainWithoutEnv(c_argc: c_int, c_argv: [*][*:0]c_char) callconv(.C) c_int {
...
 pub inline fn callMain() u8 {
-    switch (@typeInfo(@typeInfo(@TypeOf(root.main)).Fn.return_type.?)) {
-        .NoReturn => {
-            root.main();
-        },
-        .Void => {
+    const main_info = @typeInfo(@TypeOf(root.main)).Fn;
+    const ReturnType = main_info.return_type.?;
+
+    // We could return from this switch with a bunch of labeled blocks but that is ugly
```
Comment (Rexicon226) [inline lib/std/start.zig:513]: This comment serves no purpose.

Comment (gooncreeper) [reply lib/std/start.zig]: fixed

**Comment by nektro** [comment]
all the type comparisons are happening at comptime and I think the previous version of the code is more clear

**Comment by mlugg** [comment]
This code is not technically equivalent, as there may be types other than `u8` which are 8-bit unsigned integers. That said, I think the only type this could apply to is `c_char` on targets where C `char` is unsigned, and you certainly shouldn't be using that type here. So I think this is a reasonable simplification.

**Comment by mlugg** [comment]
I would also say that the changes within the `ErrorUnion` case read worse now. There is no benefit to splitting the error-handling logic into a function which is now called twice.

**Comment by gooncreeper** [comment]
> I would also say that the changes within the `ErrorUnion` case read worse now. There is no benefit to splitting the error-handling logic into a function which is now called twice.

Agree, fixed and it reads much better now.

**Pull request merged**